analyzers/src/SonarAnalyzer.CSharp/Rules/LoopsAndLinq.cs

```diff
@@ -0,0 +1,137 @@
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var forEachStatementSyntax = (ForEachStatementSyntax)c.Node;
+                    if (CanBeSimplifiedUsingWhere(forEachStatementSyntax.Statement))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, forEachStatementSyntax.Expression.GetLocation(), WhereMessageFormat));
+                    }
+                    else
+                    {
+                        CheckIfCanBeSimplifiedUsingSelect(forEachStatementSyntax, c);
+                    }
+                },
+                SyntaxKind.ForEachStatement);
+
+        private static bool CanBeSimplifiedUsingWhere(SyntaxNode statement) =>
+            GetIfStatement(statement) is { } ifStatementSyntax
+            && CanIfStatementBeMoved(ifStatementSyntax);
+
+        private static IfStatementSyntax GetIfStatement(SyntaxNode node) =>
+            node switch
+            {
+                IfStatementSyntax ifStatementSyntax => ifStatementSyntax,
+                BlockSyntax blockSyntax when blockSyntax.ChildNodes().Count() == 1 => GetIfStatement(blockSyntax.ChildNodes().Single()),
+                _ => null
+            };
+
+        private static bool CanIfStatementBeMoved(IfStatementSyntax ifStatementSyntax) =>
+            ifStatementSyntax.Else == null
+            && ifStatementSyntax.Condition is InvocationExpressionSyntax invocationExpressionSyntax
+            && !invocationExpressionSyntax.DescendantNodes()
+                                          .OfType<ArgumentSyntax>()
+                                          .Any(argument => argument.RefOrOutKeyword.IsAnyKind(SyntaxKind.OutKeyword, SyntaxKind.RefKeyword));
+
+        private static void CheckIfCanBeSimplifiedUsingSelect(ForEachStatementSyntax forEachStatementSyntax, SyntaxNodeAnalysisContext c)
+        {
+            // There are multiple scenarios where the code can be simplified using LINQ.
+            // For simplicity, in the first version of the rule, we consider that Select() can be used
+            // only when a single property from the foreach variable is used.
+            // The issue is raised if:
+            //  - the property is used more than once
+            //  - the property is the right side of a variable declaration
+            var declaredSymbol = c.SemanticModel.GetDeclaredSymbol(forEachStatementSyntax);
+            var accessedProperties = new Dictionary<ISymbol, UsageStats>();
+
+            foreach (var identifierSyntax in GetIdentifiers(forEachStatementSyntax))
+            {
+                if (identifierSyntax.Parent is MemberAccessExpressionSyntax { Parent: not InvocationExpressionSyntax } memberAccessExpressionSyntax
```

Comment:
why `Parent: not InvocationExpressionSyntax` ? maybe add a short comment?

Reply:
I think that moving method invocations can be debatable since depending on the method being called it can make the code harder to read.

I've added a test case for this since I realized it was missing.

Reply:
could you add a comment "we don't want to move method invocations" in the code? 

as a reader who didn't write the rule, it's quite difficult to find the test case which covers this without a helping comment

Reply:
What do you think about this one:
https://github.com/SonarSource/sonar-dotnet/pull/4874/files#diff-7d38d8331402fc4becf0c5d9dd102d11e36b953c2455d84f5ba20b001f39d2cbR86
We have this comment describing the big picture and it is already part of it.